sr3 stop problem

sr3 stop ~~often takes a long time to stop processes~~ can't cleanly stop instances >1 . It goes through 5 SIGTERM attempts then proceeds to kill the process.

```
sr3 stop sender/config_with_2_instances
 Stopping: sending SIGTERM .. ( 2 ) Done
 Waiting 1 sec. to check if 2 processes stopped (try: 0)
 Waiting 2 sec. to check if 2 processes stopped (try: 1)
 Waiting 4 sec. to check if 2 processes stopped (try: 2)
 Waiting 8 sec. to check if 2 processes stopped (try: 3)
 Waiting 16 sec. to check if 2 processes stopped (try: 4)
 doing SIGKILL this time
 signal_pid( 18704, SIGKILL )
 .Done
```

Pid 18704 was instance 2 for this config.

@andreleblanc11 has tried doing some debugging but hasn't really found anything obvious yet.


Log from instance 1:

```
2024-08-19 16:19:07,035 [INFO] root stop_signal signal 15 received
2024-08-19 16:19:07,035 [INFO] sarracenia.flow please_stop ok, telling 5 callbacks about it.
2024-08-19 16:19:07,327 [INFO] sarracenia.flow run starting last pass (without gather) through loop for cleanup.
2024-08-19 16:19:07,327 [INFO] sarracenia.flow _runHousekeeping on_housekeeping pid: 18703 sender/config_with_2_instances instance: 1
2024-08-19 16:19:07,327 [INFO] sarracenia.flowcb.gather.message on_housekeeping messages: good: 0 bad: 0 bytes: 0 Bytes average: 0 Bytes
2024-08-19 16:19:07,328 [INFO] sarracenia.flowcb.housekeeping.resources on_housekeeping Current cpu_times: user=14983.73 system=157.61
2024-08-19 16:19:07,328 [INFO] sarracenia.flowcb.housekeeping.resources on_housekeeping Current mem usage: 311.5 MiB, accumulating count (4 or 2/100 so far) before self-setting threshold
2024-08-19 16:19:07,328 [INFO] sarracenia.flowcb.log stats version: 3.00.53, started: 20 days ago, last_housekeeping: 256.6 seconds ago
2024-08-19 16:19:07,328 [INFO] sarracenia.flowcb.log stats messages received: 0, accepted: 0, rejected: 0   rate accepted: 0.0% or 0.0 m/s
2024-08-19 16:19:07,328 [INFO] sarracenia.flowcb.log stats files transferred: 0 bytes: 0 Bytes rate: 0 Bytes/sec
2024-08-19 16:19:07,501 [INFO] sarracenia.flow _runHousekeeping on_housekeeping pid: 18703 sender/config_with_2_instances instance: 1
2024-08-19 16:19:07,501 [INFO] sarracenia.flowcb.gather.message on_housekeeping messages: good: 0 bad: 0 bytes: 0 Bytes average: 0 Bytes
2024-08-19 16:19:07,503 [INFO] sarracenia.flowcb.housekeeping.resources on_housekeeping Current cpu_times: user=14983.9 system=157.62
2024-08-19 16:19:07,503 [INFO] sarracenia.flowcb.housekeeping.resources on_housekeeping Current mem usage: 311.5 MiB, accumulating count (4 or 2/100 so far) before self-setting threshold
2024-08-19 16:19:07,503 [INFO] sarracenia.flowcb.log stats version: 3.00.53, started: 20 days ago, last_housekeeping:  0.2 seconds ago
2024-08-19 16:19:07,503 [INFO] sarracenia.flowcb.log stats messages received: 0, accepted: 0, rejected: 0   rate accepted: 0.0% or 0.0 m/s
2024-08-19 16:19:07,504 [INFO] sarracenia.flowcb.log stats files transferred: 0 bytes: 0 Bytes rate: 0 Bytes/sec
2024-08-19 16:19:07,504 [INFO] sarracenia.flow run clean stop from run loop
2024-08-19 16:19:07,504 [INFO] sarracenia.flowcb.post.message on_stop closing
2024-08-19 16:19:07,504 [INFO] sarracenia.flowcb.gather.message on_stop closing
2024-08-19 16:19:07,504 [INFO] sarracenia.flow close flow/close completed cleanly pid: 18703 sender/config_with_2_instances instance: 1
```

Instance 2:

```
2024-08-19 16:19:07,035 [INFO] sarracenia.moth.amqp _amqp_setup_signal_handler ok, asked to stop
2024-08-19 16:19:38,586 [INFO] sarracenia.flow _runHousekeeping on_housekeeping pid: 18704 sender/hpfx-collab-wxo-dd-direct instance: 2
2024-08-19 16:19:38,586 [INFO] sarracenia.flowcb.gather.message on_housekeeping messages: good: 0 bad: 0 bytes: 0 Bytes average: 0 Bytes
2024-08-19 16:19:38,588 [INFO] sarracenia.flowcb.housekeeping.resources on_housekeeping Current cpu_times: user=20269.88 system=185.04
2024-08-19 16:19:38,588 [INFO] sarracenia.flowcb.housekeeping.resources on_housekeeping Current mem usage: 313.1 MiB, accumulating count (7 or 5/100 so far) before self-setting threshold
2024-08-19 16:19:38,588 [INFO] sarracenia.flowcb.log stats version: 3.00.53, started: 20 days ago, last_housekeeping: 304.1 seconds ago
2024-08-19 16:19:38,588 [INFO] sarracenia.flowcb.log stats messages received: 0, accepted: 0, rejected: 0   rate accepted: 0.0% or 0.0 m/s
2024-08-19 16:19:38,588 [INFO] sarracenia.flowcb.log stats files transferred: 0 bytes: 0 Bytes rate: 0 Bytes/sec
```
 

Instance 2's amqp signal handler receives the signal, but apparently not the main flow/__init__ loop?

I'm optimistic that the problem is fixed completely with this commit: af5d0ffe16df08b7c13660df8805451f9fef76db

The new interruptible_sleep function is a reusable sleep that can be interrupted by a signal.

I had to rearrange the stopping routines in `flow/__init__.py` to ensure that the please_stop callbacks are called from within the signal handler. But I did not reintroduce the infinite loop issue.

Now the moth/amqp long sleeps can be interrupted by a signal, without needing to have a signal handler inside the amqp class.


Here's an example. I ran `sr3 stop subscribe/swob` at 19:24:52, and the interruptible_sleep function returned control back to moth.amqp.getNewMessage at 19:24:55, confirming that it did not sleep for the full 64 seconds requested.

```
2024-08-20 19:22:42,855 [ERROR] sarracenia.moth.amqp getSetup AMQP getSetup failed to dd.weather.invalid.gc.ca with [Errno -2] Name or service not known
2024-08-20 19:22:42,855 [INFO] sarracenia.moth.amqp getSetup Sleeping 64 seconds ...
2024-08-20 19:23:46,914 [ERROR] sarracenia.moth.amqp getSetup connecting to: q_anonymous.subscribe.swob.sunderlandr_edcm-dirt22-4.edc-mtl.ec.gc.ca_24335702, durable: True, expire: 25200.0, auto_delete=False
2024-08-20 19:23:46,914 [ERROR] sarracenia.moth.amqp getSetup AMQP getSetup failed to dd.weather.invalid.gc.ca with [Errno -2] Name or service not known
2024-08-20 19:23:46,914 [INFO] sarracenia.moth.amqp getSetup Sleeping 64 seconds ...
2024-08-20 19:24:50,971 [ERROR] sarracenia.moth.amqp getSetup connecting to: q_anonymous.subscribe.swob.sunderlandr_edcm-dirt22-4.edc-mtl.ec.gc.ca_24335702, durable: True, expire: 25200.0, auto_delete=False
2024-08-20 19:24:50,971 [ERROR] sarracenia.moth.amqp getSetup AMQP getSetup failed to dd.weather.invalid.gc.ca with [Errno -2] Name or service not known
2024-08-20 19:24:50,971 [INFO] sarracenia.moth.amqp getSetup Sleeping 64 seconds ...
2024-08-20 19:24:53,116 [INFO] root stop_signal signal 15 received
2024-08-20 19:24:53,116 [INFO] sarracenia.flow stop_request telling 5 callbacks to please_stop.
2024-08-20 19:24:53,116 [INFO] sarracenia.flow please_stop asked to stop
2024-08-20 19:24:53,117 [DEBUG] sarracenia.flowcb.gather.message please_stop asking Moth consumer to please_stop
2024-08-20 19:24:53,117 [INFO] sarracenia.moth please_stop asked to stop
2024-08-20 19:24:55,974 [WARNING] sarracenia.moth.amqp getNewMessage failed q_anonymous.subscribe.swob.sunderlandr_edcm-dirt22-4.edc-mtl.ec.gc.ca_24335702: 'AMQP' object has no attribute 'channel'
2024-08-20 19:24:55,974 [WARNING] sarracenia.moth.amqp getNewMessage lost connection to broker
2024-08-20 19:24:56,975 [DEBUG] sarracenia.flow filter start len(incoming)=0, rejected=0
2024-08-20 19:24:56,976 [DEBUG] sarracenia.flow filter end len(incoming)=0, rejected=0
2024-08-20 19:24:56,976 [DEBUG] sarracenia.flow filter B filtered incoming: 0, ok: 0 (directories: 0), rejected: 0, failed: 0 stop_requested: True have_vip: ['AnyAddressIsFine']
2024-08-20 19:24:56,976 [DEBUG] sarracenia.flow do processing 0 messages worked!

...

2024-08-20 19:24:56,996 [DEBUG] sarracenia.flow run  not throttling: limit: 0
2024-08-20 19:24:56,997 [DEBUG] sarracenia.flow run clean stop from run loop
2024-08-20 19:24:56,997 [INFO] sarracenia.flowcb.gather.message on_stop closing
2024-08-20 19:24:56,997 [INFO] sarracenia.flow close flow/close completed cleanly pid: 1931974 subscribe/swob instance: 1
```

The whole instance shut down cleanly shortly after that.

--- a/sarracenia/flow/__init__.py
+++ b/sarracenia/flow/__init__.py
@@ -461,7 +461,7 @@ class Flow:
         m.setReport(code, reason)
 
     def please_stop(self) -> None:
-        logger.info( f'ok, telling {len(self.plugins["please_stop"])} callbacks about it.')
+        logger.info(f'stop requested')
         self._stop_requested = True
         self.metrics["flow"]['stop_requested'] = True
 
@@ -562,8 +562,10 @@ class Flow:
                     self.close()
                     break
                 else:
-                    logger.debug( 'starting last pass (without gather) through loop for cleanup.')
+                    logger.debug('starting last pass (without gather) through loop for cleanup.')
                     stopping = True
+                    logger.info(f'telling {len(self.plugins["please_stop"])} callbacks to please_stop.')
+                    self.runCallbacksTime('please_stop')
 
             self._run_vip_update()
 

--- a/sarracenia/flow/__init__.py
+++ b/sarracenia/flow/__init__.py
@@ -613,7 +613,9 @@ class Flow:
             run_time = now - start_time
             total_messages += last_gather_len
 
+            # trigger shutdown when messageCountMax is reached
             if (self.o.messageCountMax > 0) and (total_messages > self.o.messageCountMax):
+                logger.info(f'{total_messages} messages processed > messageCountMax {self.o.messageCountMax}')
                 self.runCallbacksTime('please_stop')
 
             current_rate = total_messages / run_time
@@ -622,6 +624,7 @@ class Flow:
             self.metrics['flow']['msgRate'] = current_rate
             self.metrics['flow']['msgRateCpu'] = total_messages / (self.metrics['flow']['cpuTime']+self.metrics['flow']['last_housekeeping_cpuTime'] )
 
+            # trigger shutdown once gather is finished, where sleep < 0 (e.g. a post)
             if (last_gather_len == 0) and (self.o.sleep < 0):
                 if (self.o.retryEmptyBeforeExit and "retry" in self.metrics
                     and self.metrics['retry']['msgs_in_post_retry'] > 0):

--- a/sarracenia/flow/__init__.py
+++ b/sarracenia/flow/__init__.py
@@ -460,12 +460,21 @@ class Flow:
         self.worklist.rejected.append(m)
         m.setReport(code, reason)
 
+    def stop_request(self) -> None:
+        """ called by the signal handler to tell self and FlowCB classes to stop. Without this,
+            calling runCallbacksTime('please_stop') from inside self.please_stop causes an infinite loop.
+            Note: if this is called from outside of a signal handler, the interruptible_sleep function
+                  won't work.
+        """
+        logger.info(f'telling {len(self.plugins["please_stop"])} callbacks to please_stop.')
+        # this will call the please_stop method below, and other classes' please_stop methods
+        self.runCallbacksTime('please_stop')
+
     def please_stop(self) -> None:
-        logger.info(f'stop requested')
+        logger.info(f'asked to stop')
         self._stop_requested = True
         self.metrics["flow"]['stop_requested'] = True
-
-
+    
     def close(self) -> None:
 
         self.runCallbacksTime('on_stop')
@@ -564,8 +573,6 @@ class Flow:
                 else:
                     logger.debug('starting last pass (without gather) through loop for cleanup.')
                     stopping = True
-                    logger.info(f'telling {len(self.plugins["please_stop"])} callbacks to please_stop.')
-                    self.runCallbacksTime('please_stop')
 
             self._run_vip_update()
 

--- a/sarracenia/flow/__init__.py
+++ b/sarracenia/flow/__init__.py
@@ -1008,7 +1008,7 @@ class Flow:
 
                 if self.o.fileAgeMin > 0 and age < self.o.fileAgeMin:
                     logger.warning( f"file too young: queueing for retry.")
-                    self.worklist.failed.append(msg)
+                    self.worklist.failed.append(m)
                     continue
 
             if 'fileOp' in m and 'rename' in m['fileOp']:

--- a/sarracenia/flowcb/gather/message.py
+++ b/sarracenia/flowcb/gather/message.py
@@ -74,3 +74,10 @@ class Message(FlowCB):
         if hasattr(self,'consumer') and hasattr(self.consumer, 'close'):
             self.consumer.close()
         logger.info('closing')
+
+    def please_stop(self) -> None:
+        """ pass stop request along to consumer Moth instance(s)
+        """
+        super().please_stop()
+        if hasattr(self,'consumer') and self.consumer:
+            self.consumer.please_stop()

--- a/sarracenia/flowcb/gather/message.py
+++ b/sarracenia/flowcb/gather/message.py
@@ -80,4 +80,5 @@ class Message(FlowCB):
         """
         super().please_stop()
         if hasattr(self,'consumer') and self.consumer:
+            logger.debug("asking Moth consumer to please_stop")
             self.consumer.please_stop()

--- a/sarracenia/flowcb/post/message.py
+++ b/sarracenia/flowcb/post/message.py
@@ -69,3 +69,10 @@ class Message(FlowCB):
         if hasattr(self,'poster') and self.poster:
             self.poster.close()
         logger.debug('closing')
+    
+    def please_stop(self) -> None:
+        """ pass stop request along to publisher Moth instance(s)
+        """
+        super().please_stop()
+        if hasattr(self, 'poster') and self.poster:
+            self.poster.please_stop()

--- a/sarracenia/flowcb/post/message.py
+++ b/sarracenia/flowcb/post/message.py
@@ -75,4 +75,5 @@ class Message(FlowCB):
         """
         super().please_stop()
         if hasattr(self, 'poster') and self.poster:
+            logger.debug("asking Moth publisher to please_stop")
             self.poster.please_stop()

--- a/sarracenia/flowcb/report.py
+++ b/sarracenia/flowcb/report.py
@@ -166,3 +166,9 @@ class Report(FlowCB):
             self.poster.close()
         logger.info('closing')
 
+    def please_stop(self) -> None:
+        """ pass stop request along to publisher Moth instance(s)
+        """
+        super().please_stop()
+        if hasattr(self, 'poster') and self.poster:
+            self.poster.please_stop()

--- a/sarracenia/flowcb/report.py
+++ b/sarracenia/flowcb/report.py
@@ -171,4 +171,5 @@ class Report(FlowCB):
         """
         super().please_stop()
         if hasattr(self, 'poster') and self.poster:
+            logger.debug("asking Moth publisher to please_stop")
             self.poster.please_stop()

--- a/sarracenia/instance.py
+++ b/sarracenia/instance.py
@@ -62,7 +62,7 @@ class instance:
                     if line:
                         code.append("  %s" % (line.strip()))
             logging.debug('\n'.join(code))
-        self.running_instance.please_stop()
+        self.running_instance.stop_request()
 
     def start(self):
         """

--- a/sarracenia/interruptible_sleep.py
+++ b/sarracenia/interruptible_sleep.py
@@ -1,0 +1,32 @@
+"""
+Long running sleeps prevent Sarracenia from being shutdown cleanly in a
+reasonable amount of time.
+
+This implements a reusable sleep function that can be called from other parts
+of the code to sleep for a long time, but can still be interrupted.
+"""
+
+import time
+
+def interruptible_sleep(sleep_time:float, obj: object, stop_flag_name: str='_stop_requested', nap_time: float=5.0) -> bool:
+    """ Sleep for sleep_time, divided up into shorter nap_time intervals.
+        Pass a reference to an object that contains a boolean attribute named stop_flag_name. 
+        Between each nap_time, the function will check if obj.stop_flag_name has become True.
+        If the flag is False, it will continue sleeping, if True, it will abort the sleep.
+
+         Args:
+            sleep_time (float): total amount of time to sleep, if not interrupted
+            obj (object): the object containing the boolean attribute named stop_flag_name
+            stop_flag_name (str): the name of the boolean attribute in obj
+            nap_time (float): default = 5.0, sleep in intervals of nap_time
+
+        Returns:
+            bool: ``True`` if the sleep **was** interrupted, ``False`` if it slept for the entire ``sleep_time`` 
+                    time without interruption.
+    """
+    interrupted = False
+    while sleep_time > 0 and not interrupted:
+        time.sleep(min(nap_time, sleep_time))
+        sleep_time -= nap_time
+        interrupted = (hasattr(obj, stop_flag_name) and getattr(obj, stop_flag_name))
+    return interrupted

--- a/sarracenia/moth/__init__.py
+++ b/sarracenia/moth/__init__.py
@@ -280,7 +280,7 @@ class Moth():
 
         self.is_subscriber = is_subscriber
         self.connected=False
-        self.please_stop = False
+        self.self._stop_requested = False
         self.metrics = { 'connected': False }
         self.metricsReset()
 
@@ -349,6 +349,13 @@ class Moth():
         """
         logger.error("NewMessages unimplemented")
         return []
+    
+    def please_stop(self) -> None:
+        """ register a request to cleanly stop. Any long running processes should check for _stop_requested and 
+            stop if it becomes True.
+        """
+        logger.info("asked to stop")
+        self._stop_requested = True
 
     def putNewMessage(self, message:sarracenia.Message, content_type: str ='application/json', exchange: str = None) -> bool:
         """

--- a/sarracenia/moth/__init__.py
+++ b/sarracenia/moth/__init__.py
@@ -280,7 +280,7 @@ class Moth():
 
         self.is_subscriber = is_subscriber
         self.connected=False
-        self.self._stop_requested = False
+        self._stop_requested = False
         self.metrics = { 'connected': False }
         self.metricsReset()
 

--- a/sarracenia/moth/amqp.py
+++ b/sarracenia/moth/amqp.py
@@ -34,7 +34,6 @@ import re
 import sarracenia
 from sarracenia.postformat import PostFormat
 from sarracenia.moth import Moth
-import signal
 import os
 
 import time
@@ -182,7 +181,7 @@ class AMQP(Moth):
         self.o.update(props)
 
         self.first_setup = True
-        self.please_stop = False
+        self._stop_requested = False
 
         me = "%s.%s" % (__class__.__module__, __class__.__name__)
 
@@ -241,10 +240,6 @@ class AMQP(Moth):
         self.management_channel = self.connection.channel(1)
         self.channel = self.connection.channel(2)
         return True
-
-    def _amqp_setup_signal_handler(self, signum, stack):
-        logger.info("ok, asked to stop")
-        self.please_stop=True
 
     def metricsReport(self):
 
@@ -327,14 +322,10 @@ class AMQP(Moth):
              connect, declare queue, apply bindings.
         """
         ebo = 1
-        original_sigint = signal.getsignal(signal.SIGINT)
-        original_sigterm = signal.getsignal(signal.SIGINT)
-        signal.signal(signal.SIGINT, self._amqp_setup_signal_handler)
-        signal.signal(signal.SIGTERM, self._amqp_setup_signal_handler)
 
         while True:
             
-            if self.please_stop:
+            if self._stop_requested:
                 break
 
             if 'broker' not in self.o or self.o['broker'] is None:
@@ -402,25 +393,16 @@ class AMQP(Moth):
             logger.info("Sleeping {} seconds ...".format(ebo))
             time.sleep(ebo)
 
-        signal.signal(signal.SIGINT, original_sigint)
-        signal.signal(signal.SIGTERM, original_sigterm)
-        if self.please_stop:
-            os.kill(os.getpid(), signal.SIGINT)
-
     def putSetup(self) -> None:
 
         ebo = 1
-        original_sigint = signal.getsignal(signal.SIGINT)
-        original_sigterm = signal.getsignal(signal.SIGINT)
-        signal.signal(signal.SIGINT, self._amqp_setup_signal_handler)
-        signal.signal(signal.SIGTERM, self._amqp_setup_signal_handler)
 
         while True:
 
             # It does not really matter how it fails, the recovery approach is always the same:
             # tear the whole thing down, and start over.
             try:
-                if self.please_stop:
+                if self._stop_requested:
                     break
 
                 if self.o['broker'] is None:
@@ -475,12 +457,6 @@ class AMQP(Moth):
             self.close()
             logger.info("Sleeping {} seconds ...".format(ebo))
             time.sleep(ebo)
-
-        signal.signal(signal.SIGINT, original_sigint)
-        signal.signal(signal.SIGTERM, original_sigterm)
-        if self.please_stop:
-            os.kill(os.getpid(), signal.SIGINT)
-
 
     def putCleanUp(self) -> None:
 

--- a/sarracenia/moth/amqp.py
+++ b/sarracenia/moth/amqp.py
@@ -34,6 +34,7 @@ import re
 import sarracenia
 from sarracenia.postformat import PostFormat
 from sarracenia.moth import Moth
+from sarracenia.interruptible_sleep import interruptible_sleep
 import os
 
 import time
@@ -391,7 +392,7 @@ class AMQP(Moth):
             if ebo < 60: ebo *= 2
 
             logger.info("Sleeping {} seconds ...".format(ebo))
-            time.sleep(ebo)
+            interruptible_sleep(ebo, obj=self)
 
     def putSetup(self) -> None:
 
@@ -456,7 +457,7 @@ class AMQP(Moth):
 
             self.close()
             logger.info("Sleeping {} seconds ...".format(ebo))
-            time.sleep(ebo)
+            interruptible_sleep(ebo, obj=self)
 
     def putCleanUp(self) -> None:
 
@@ -612,7 +613,7 @@ class AMQP(Moth):
             if ebo < 60:
                 ebo *= 2
             logger.info("Sleeping {} seconds before re-trying ack...".format(ebo))
-            time.sleep(ebo)
+            interruptible_sleep(ebo, obj=self)
             # TODO maybe implement message strategy stubborn here and give up after retrying?
 
     def putNewMessage(self,

--- a/sarracenia/moth/mqtt.py
+++ b/sarracenia/moth/mqtt.py
@@ -33,7 +33,6 @@ import paho.mqtt.client
 import sarracenia
 from sarracenia.postformat import PostFormat
 from sarracenia.moth import Moth
-import signal
 import os
 import ssl
 import threading
@@ -342,25 +341,17 @@ class MQTT(Moth):
                                unquote(self.o['broker'].url.password))
         return client
 
-    def _mqtt_setup_signal_handler(self, signum, stack):
-        logger.info("ok, asked to stop")
-        self.please_stop=True
-
     def getSetup(self):
         """
            Establish a connection to consume messages with.  
         """
         ebo = 1
-        original_sigint = signal.getsignal(signal.SIGINT)
-        original_sigterm = signal.getsignal(signal.SIGINT)
-        signal.signal(signal.SIGINT, self._mqtt_setup_signal_handler)
-        signal.signal(signal.SIGTERM, self._mqtt_setup_signal_handler)
 
         something_broke = True
         self.connected=False
         while True:
 
-            if self.please_stop:
+            if self._stop_requested:
                 break
 
             try:
@@ -390,7 +381,7 @@ class MQTT(Moth):
                     while (self.connect_in_progress) or (self.subscribe_in_progress > 0):
                          logger.info( f"waiting for subscription to be set up. (ebo={ebo})")
                          time.sleep(0.1*ebo)
-                         if self.please_stop:
+                         if self._stop_requested:
                               break
                          if ebo < 512 :
                             ebo *= 2
@@ -411,7 +402,7 @@ class MQTT(Moth):
                         while (self.connect_in_progress) or (self.subscribe_in_progress > 0):
                             logger.info( f"waiting ({ebo} seconds) for broker to confirm subscription is set up.")
                             logger.info( f"for {icid} connect_in_progress={self.connect_in_progress} subscribe_in_progress={self.subscribe_in_progress}" )
-                            if self.please_stop:
+                            if self._stop_requested:
                                 break
                             if ebo < 60: ebo *= 2
                             decl_client.loop(ebo)
@@ -427,27 +418,16 @@ class MQTT(Moth):
             if ebo < 60: ebo *= 2
             time.sleep(ebo)
 
-        signal.signal(signal.SIGINT, original_sigint)
-        signal.signal(signal.SIGTERM, original_sigterm)
-        if self.please_stop:
-            os.kill(os.getpid(), signal.SIGINT)
-
-
-
     def putSetup(self):
         """
            establish a connection to allow publishing. 
         """
         ebo = 1
-        original_sigint = signal.getsignal(signal.SIGINT)
-        original_sigterm = signal.getsignal(signal.SIGINT)
-        signal.signal(signal.SIGINT, self._mqtt_setup_signal_handler)
-        signal.signal(signal.SIGTERM, self._mqtt_setup_signal_handler)
 
         self.connected=False
         while True:
             
-            if self.please_stop:
+            if self._stop_requested:
                 break
 
             try:
@@ -489,7 +469,7 @@ class MQTT(Moth):
                 while self.connect_in_progress:
                      logger.info( f"waiting for connection to {self.o['broker']} ebo={ebo}")
                      time.sleep(0.1*ebo)
-                     if self.please_stop:
+                     if self._stop_requested:
                           break
                      if ebo < 512:
                           ebo *= 2
@@ -503,12 +483,6 @@ class MQTT(Moth):
 
             if ebo < 60: ebo *= 2
             time.sleep(ebo)
-
-        signal.signal(signal.SIGINT, original_sigint)
-        signal.signal(signal.SIGTERM, original_sigterm)
-        if self.please_stop:
-            os.kill(os.getpid(), signal.SIGINT)
-
 
     def __sub_on_message(client, userdata, msg):
         """
@@ -544,7 +518,7 @@ class MQTT(Moth):
                    clean_start=False, properties=props )
                 while self.connect_in_progress:
                     myclient.loop(0.1)
-                    if self.please_stop:
+                    if self._stop_requested:
                        break
                 myclient.disconnect()
                 logger.info( f"instance deletion for {i:02d} done" )

--- a/tests/sarracenia/interruptible_sleep_test.py
+++ b/tests/sarracenia/interruptible_sleep_test.py
@@ -1,0 +1,68 @@
+import pytest
+from tests.conftest import *
+#from unittest.mock import Mock
+
+import datetime
+import os
+import signal
+import subprocess
+from sarracenia.interruptible_sleep import interruptible_sleep
+
+class SleepThing():
+    def __init__(self):
+        self._stop_requested = False
+        self.other_name = False
+        signal.signal(signal.SIGTERM, self.signal_handler)
+        signal.signal(signal.SIGINT, self.signal_handler)
+    
+    def signal_handler(self, signum, stack):
+        self._stop_requested = True
+        self.other_name = True
+
+def test_interruptible_sleep():
+    st = SleepThing()
+    stime = 10
+
+    # Test that sleep sleeps for the right amount of time when not interrupted
+    before_time = datetime.datetime.now()
+    result = interruptible_sleep(stime, st)
+    after_time = datetime.datetime.now()
+    assert (result == False)
+    assert ( (after_time - before_time).seconds == stime)
+
+    # Test that the sleep behaves correctly when interrupted
+    # send a SIGINT to this process after 5 seconds:
+    cmdline = f"""bash -c '/usr/bin/sleep 5; kill -SIGTERM {os.getpid()};' &"""
+    subprocess.run(cmdline, shell=True)
+    before_time = datetime.datetime.now()
+    result = interruptible_sleep(stime, st)
+    after_time = datetime.datetime.now()
+    assert result
+    assert ( (after_time - before_time).seconds == 5)
+
+    # Test using a different nap_time
+    st = SleepThing()
+    # send a SIGINT to this process after 5 seconds:
+    cmdline = f"""bash -c '/usr/bin/sleep 5; kill -SIGTERM {os.getpid()};' &"""
+    subprocess.run(cmdline, shell=True)
+    before_time = datetime.datetime.now()
+    result = interruptible_sleep(stime, st, nap_time=1)
+    after_time = datetime.datetime.now()
+    assert result
+    assert ( (after_time - before_time).seconds == 5)
+
+
+    # Test using a different attribute name
+    st = SleepThing()
+    # send a SIGINT to this process after 5 seconds:
+    cmdline = f"""bash -c '/usr/bin/sleep 5; kill -SIGTERM {os.getpid()};' &"""
+    subprocess.run(cmdline, shell=True)
+    before_time = datetime.datetime.now()
+    result = interruptible_sleep(stime, st, stop_flag_name = 'other_name')
+    after_time = datetime.datetime.now()
+    assert result
+    assert ( (after_time - before_time).seconds == 5)
+
+
+
+

